File: Dynamic_Programming/629. K Inverse Pairs Array.cpp
/*
629. K Inverse Pairs Array
*/

// 數學優化 TC:O(n^2) SC:O(n^2)
class Solution {
public:
    int kInversePairs(int n, int k) {
        
        vector<vector<long>> dp(n+1, vector<long>(k+1,0));
        for(int i = 0; i <= n; i++){
            dp[i][0] = 1;
        } 

        int M = 1e9+7;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= k; j++){
                if(j >= i)
                    dp[i][j] = (dp[i][j-1] + dp[i-1][j] - dp[i-1][j-i] + M) % M;
                else dp[i][j] = (dp[i][j-1] + dp[i-1][j]) % M;
               
            }
        }

        return dp[n][k];
    }
};


// TC:O(n^3) SC:O(n^2)
class Solution {
public:
    int kInversePairs(int n, int k) {
        
        vector<vector<int>> dp(n+1, vector<int>(k+1,0));
        for(int i = 0; i <= n; i++){
            dp[i][0] = 1;
        } 

        int M = 1e9+7;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= k; j++){
                for(int m = 0; m <= min(j,i-1); m++){
                    dp[i][j] += dp[i-1][j-m];
                    dp[i][j] %= M;
                }
            }
        }

        return dp[n][k];
    }
};

/*
給你一個整數計算這個整數的排列組合中有多少滿足以下 並且給定要的對數k
inverse pair -> i < j , nums[i] > nums[j]
dp[i][對數] : 加入數字i時k=j時有多少種放法
[x x x x x] 6
看能插入哪裡 dp[6][j] += dp[5][j-5](插在頭) + dp[5][j-4](插在第二個)....
dp[i][j] += sum(dp[i-1][j-k]) for k = 0,....i-1 
這樣寫其實就能過但是n^3過得很免強 
可以再用數學來優化
dp[i][j] = dp[i-1][j-0] + dp[i-1][j-1] + ... dp[i-1][j-(i-1)];
dp[i][j-1] = dp[i-1][j-1] + dp[i-1][j-2] + ... dp[i-1][j-i];
dp[i][j] - dp[i][j-1] = dp[i-1][j] - dp[i-1][j-i];
dp[i][j] = dp[i][j-1] + dp[i-1][j] - dp[i-1][j-i];
*/

File: Dynamic_Programming/629. K Inverse Pairs Array.py
# TC:O(n * k) SC:O(n * k)
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        '''
        這題要求剛好要有k個逆序對 total 問說會有多少種組合方式?
        1 ~ n 的數字
        這題如果像3193 寫一個n^3 是不行的， 要優化到n^2
        '''
        MOD = 10 ** 9 + 7
        # 空間優化
        f = [0] * (k + 1)
        f[0] = 1
        for i in range(1, n):
            pre = list(accumulate(f, initial=0))
            new_f = [0] * (k + 1)
            for j in range(min(k, i * (i + 1) // 2) + 1): # 當下可以組成多少逆序對
                mn = max(0, j - i)
                new_f[j] = (pre[j+1] - pre[mn]) % MOD
                # for t in range(0, i+1):
                #     if j >= t:
                #         f[i][j] = (f[i][j] + f[i-1][j - t]) % MOD
            f = new_f.copy()
        return f[k]
        '''
        f = [[0] * (k + 1) for _ in range(n)]
        f[0][0] = 1
        for i in range(1, n):
            pre = list(accumulate(f[i-1], initial=0))
            for j in range(min(k, i * (i + 1) // 2) + 1): # 當下可以組成多少逆序對
                mn = max(0, j - i)
                f[i][j] = (pre[j+1] - pre[mn]) % MOD
                # for t in range(0, i+1):
                #     if j >= t:
                #         f[i][j] = (f[i][j] + f[i-1][j - t]) % MOD
        return f[n-1][k]
        '''


        '''
        遞歸TLE
        @cache
        def dfs(i:int, cnt:int) -> int:
            if i == 0:
                return 1 if cnt == 0 else 0
            ans = 0
            for j in range(i + 1):
                ans = (ans + dfs(i - 1, cnt - j)) % MOD
            return ans
        return dfs(n-1, k)
        '''
